File: typescript/src/tools/insights/update.ts
import { getProjectBaseUrl } from "@/lib/utils/api";
import { InsightUpdateSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = InsightUpdateSchema;

type Params = z.infer<typeof schema>;

export const updateHandler = async (context: Context, params: Params) => {
	const { insightId, data } = params;
	const projectId = await context.getProjectId();
	const insightResult = await context.api.insights({ projectId }).update({
		insightId,
		data,
	});

	if (!insightResult.success) {
		throw new Error(`Failed to update insight: ${insightResult.error.message}`);
	}

	const insightWithUrl = {
		...insightResult.data,
		url: `${getProjectBaseUrl(projectId)}/insights/${insightResult.data.short_id}`,
	};

	return { content: [{ type: "text", text: JSON.stringify(insightWithUrl) }] };
};

const definition = getToolDefinition("insight-update");

const tool = (): Tool<typeof schema> => ({
	name: "insight-update",
	description: definition.description,
	schema,
	handler: updateHandler,
});

export default tool;

File: python/tools/insights/update.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.tool_inputs import InsightUpdateSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def update_insight_handler(context: Context, params: InsightUpdateSchema) -> ToolResult:
    project_id = await context.get_project_id()
    insight_result = await context.api.insights(project_id).update(int(params.insightId), params.data)

    if is_error(insight_result):
        raise Exception(f"Failed to update insight: {insight_result.error}")

    assert is_success(insight_result)

    insight_data = insight_result.data

    insight_with_url = {
        **insight_data.model_dump(),
        "url": f"{get_project_base_url(project_id, context.api.base_url)}/insights/{insight_data.short_id}",
    }

    return ToolResult(content=json.dumps(insight_with_url))


def update_insight_tool() -> Tool[InsightUpdateSchema]:
    definition = get_tool_definition("insight-update")

    return Tool(
        name="insight-update",
        description=definition["description"],
        schema=InsightUpdateSchema,
        handler=update_insight_handler,
    )
